### PR TITLE
Improved GMLAN support. Minor CAN bus bug fixes and improvements

### PR DESCRIPTION
**Pull request by mjuhanne**
* Allow user to set GMLAN Single Wire CAN operation modes (Normal, High Voltage Wake Up, High Speed, Sleep)
* Change SJW=2 for GMLAN Single Wire CAN
* Add non-blocking CAN read function
* Log can_send_errs / gmland_send_errs properly
* Fix lost_rx errors counted as tx_errors
* More verbose logging in case of CAN tx/arbitration lost error

### CODE REVIEW

**Comment by mjuhanne** [comment]
Ok, is this a false positive warning by MISRA? 
```
Failed! found Misra violations in panda code:
[board/drivers/can.h:377] (style) misra violation (use --rule-texts=<file> to get proper output) (Undefined) [misra-c2012-15.7]
```
_**Misra c 2012 15.7: All if . . else if constructs shall be terminated with an else statement**_

But there is no "else if" statement there...

File: board/drivers/can.h [hunk 1]
```diff
@@ -387,16 +463,21 @@ void CAN3_RX0_IRQHandler(void) { can_rx(2); }
...
+  bool res;
   if (safety_tx_hook(to_push) != 0) {
     if (bus_number < BUS_MAX) {
       // add CAN packet to send queue
       // bus number isn't passed through
       to_push->RDTR &= 0xF;
       if ((bus_number == 3U) && (can_num_lookup[3] == 0xFFU)) {
         // TODO: why uint8 bro? only int8?
-        gmlan_send_errs += !bitbang_gmlan(to_push);
+        res = bitbang_gmlan(to_push);
```
Comment (geohot) [inline board/drivers/can.h]: What was wrong with += before?

**Comment by geohot** [comment]
The MISRA thing is not a false positive. If you want to use an if statement, you need to handle the else case as well.

File: board/drivers/can.h (hunk 1 above)
Comment (rbiasini) [reply board/drivers/can.h]: Bool added to a int. Misra does not like it. It's already fixed in master, just need to resolve conflicts.

File: board/drivers/can.h [hunk 2]
```diff
@@ -326,6 +346,60 @@ void process_can(uint8_t can_number) {
         CAN->sTxMailBox[0].TDTR = to_send.RDTR;
         CAN->sTxMailBox[0].TIR = to_send.RIR;
       }
+    } else {
+      if ((CAN->TSR & CAN_TSR_TERR0) == CAN_TSR_TERR0) {
```
Comment (rbiasini) [inline board/drivers/can.h]: This is the same as "else if", which is why misra complains.

**Comment by geohot** [comment]
If you still want this, please rebase and resubmit

**Pull request closed without merging**